### PR TITLE
Deprecate copy methods on configuration

### PR DESCRIPTION
**Pull request by tresat**
The 3 failing checks appear to be transitory and irrelevant, this is ready for review.

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
No DeprecationLogger? No upgrade guide section?

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
> I just wanted to see where smoke tests failed

I know the [gradle versions plugin](https://github.com/ben-manes/gradle-versions-plugin) relies on this functionality heavily. We probably need to come up with an alternative implementation for that plugin or otherwise introduce an API which can support its use case. 

The maintainer [has described](https://github.com/ben-manes/gradle-versions-plugin/issues/718#issuecomment-1372547651) why they copy configurations:
```
We copy the configurations so that we can rewrite the dependencies to use dynamic versions. We then resolve the original and our modified version, compare the versions selected, and report on this difference. [Originally](https://discuss.gradle.org/t/add-report-to-display-dependency-updates/540) this was a detached configuration, but there were too many Gradle bugs (like repository handler's deduplication logic either adding multiple or dropping repositories), so copying was the safest option. It also meant that we could honor configuration that was not always exposed, like resolution strategies.
```

In short, they were using detached configurations, but in reality they wanted to do resolution in the same context as the original resolvable configuration but with different dependencies. For example, the plugin copies the `runtimeClasspath` configuration but adds different dependencies to it and resolves the copy. This would not be possible with a detached configuration since the detached configuration has separate attributes, resolution strategy, etc. compared to the resolution context they wanted to resolve in. 

I think a longer term solution to this would be some public `ResolveContext` API similar to how we're introducing a separate API for variants which is independent of configurations. Short term we may need to think of some other options

**Comment by tresat** [comment]
> I know the [gradle versions plugin](https://github.com/ben-manes/gradle-versions-plugin) relies on this functionality heavily. We probably need to come up with an alternative implementation for that plugin or otherwise introduce an API which can support its use case.

Why won't extend and modify the extended configuration work as a replacement for copy?  Attributes, resolutionStrategies, etc., can be copied individually.  Alternately, with just a little more work an entirely new configuration can be created.  It looks like the only internal API used is `getAllExcludeRules()` which looks like it could be rewritten using the public APIs pretty easily.  I think this is definitely fat that could be trimmed from our implementation, and users who really want this functionality can reimplement it.

There are already a few expected deprecation warnings thrown by the versions plugin in our smoke tests, so I don't see a few more making any substantial difference.  A deprecation warning isn't something that needs to be avoided at all costs, it's a hint that something is not using current best practices and will have problems in the future.  This seems true in this case whether we alert users by providing the message or not.  And if we deprecate and get a lot of feedback that people must have the ability to copy we can:
 1) delay removal
 2) undeprecate it (and fix copying)
 3) provide a helper somewhere outside of the core `Configuration` type.
But the best way to get that feedback is to deprecate first.

I think the most important questions here are: 
 1) "Does it make conceptual sense to want to copy a Configuration?"  Going role by role which should we be copying rather than extending?  And why not start with a fresh one if you're going to be modifying it anyway (and if you're not, than why did you make a copy)?
 2) "What is the definition of a copy?"  Why does it copy extended exclude rules but not dependencies in superconfigurations?  This seems kind of arbitrary and surprising.  Why does the copy appear in a different state than than the original?  What other meaningful state is **not** copied (a lot, I think).  Why do we provide a core API to copy while filtering dependencies but not modifying anything else?

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
I wonder if we should consider putting a "copy" of the existing copy methods, rewritten as static helpers using only public API, in the docs.  Then the deprecation message can point users to these implementations as replacements, perhaps after making these same points about why copying isn't a good idea first.

Then submitting PRs to any external plugins to remove the deprecated usage would be trivial.

File: subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java [hunk 1]
```diff
@@ -480,6 +481,7 @@ public String determineName(Configuration c) {
      * {@link #getHierarchy()} for the copy will not include any superconfigurations.
      * @return copy of this configuration
      */
+    @Deprecated
     Configuration copyRecursive();
```
Comment (big-guy) [inline subprojects/core-api/src/main/java/org/gradle/api/artifacts/Configuration.java:485]: Do you need `@Deprecated` on the Spec-taking ones too?

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy [hunk 2]
```diff
@@ -58,14 +58,26 @@ class GradleVersionsPluginSmokeTest extends AbstractPluginValidatingSmokeTest {
...
+        copiedConfigurations.each { conf ->
+            ['sub1', 'sub2'].each { sub ->
+                runner.expectDeprecationWarning(
+                        "Copying configurations has been deprecated. This is scheduled to be removed in Gradle 9.0. Consider creating a new configuration and extending configuration ':$sub:$conf' instead. Consult the upgrading guide for further information: https://docs.gradle.org/${GradleVersion.current().version}/userguide/upgrading_version_8.html#configuration_copying_deprecated",
+                        "https://github.com/ben-manes/gradle-versions-plugin/issues/718")
+            }
+        }
+        runner.expectDeprecationWarning(
+                "Copying configurations has been deprecated. This is scheduled to be removed in Gradle 9.0. Consider creating a new configuration and extending configuration ':classpath' instead. Consult the upgrading guide for further information: https://docs.gradle.org/${GradleVersion.current().version}/userguide/upgrading_version_8.html#configuration_copying_deprecated",
```
Comment (big-guy) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy:78]: What's our recommendation for their use case here?

Comment (tresat) [reply subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy:78]: Do you mean for `classpath` specifically?  Will the extension strategy not work?

Comment (tresat) [reply subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy:78]: @big-guy - I added some documentation with an example alternative, and a sample test showing an extension made equal to the result of a copy.  The public API is limited in some things, see the test.  Would something like this be enough to go ahead with this?  I can't see why any affected plugins couldn't transition to using (more verbose) code like this to achieve whatever they were relying on `copy` to do.

Should the different pieces of copy (artifacts, attributes, etc.) be added to a public duplication helper API?

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
Something else to consider as evidence this should be deprecated:

In `release`, currently this integration test:

```
    def "copying a configuration causes extraneous warnings with config cache"() {
        buildFile.text = """  
            plugins {  
                id 'java-library'  
            }  
      
            def implementationCopy = configurations.implementation.copy()  
              
            task checkCopy {  
                doLast {  
                    configurations.implementation.allArtifacts
                    implementationCopy.allArtifacts  
                }  
            }  
        """

        expect:
        succeeds ":checkCopy"
    }
```

Will produce the warning:

`The implementationCopy configuration has been deprecated for resolution. This will fail with an error in Gradle 9.0. Please resolve another configuration instead. Consult the upgrading guide for further information: https://docs.gradle.org/8.1-20230328040000+0000/userguide/upgrading_version_5.html#dependencies_should_no_longer_be_declared_using_the_compile_and_runtime_configurations`

**only when run with the `configCache` runner**, there is no such warning for the `embedded` runner.  This behavior difference predates these changes and is the reason for ignoring deprecations in the test in the last commit here.

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy (hunk 2 above)
Comment (big-guy) [reply subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy:78]: I think manually copying things is still problematic in the same ways as `copy` is.  It's only a point-in-time copy.

I wasn't specifically thinking about `classpath`, but more the general issue and recommendation we're making here.

I think it's OK that we're saying you shouldn't use `copy`, but if we ask the versions plugin to use `extendsFrom`, I fear we're setting them up for another change in behavior. Haven't we talked about limiting which configurations can be "extended from"?  

What breaks if we also copy the intended role into the copy?

**Review by big-guy — CHANGES_REQUESTED** [review]

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy (hunk 2 above)
Comment (tresat) [reply subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy:78]: Discuses last week, working on changes.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
Won't Do.  Replaced by https://github.com/gradle/gradle/pull/24804
